utils/fixed_decimal/src/decimal.rs

```diff
@@ -989,43 +1009,190 @@ impl FixedDecimal {
+        let past_required_digits = match increment {
+            RoundingIncrement::MultiplesOf1 => {
+                // We can early exit if `position` is the last digit
+                position <= self.nonzero_magnitude_end()
+            }
+            RoundingIncrement::MultiplesOf2 | RoundingIncrement::MultiplesOf5 => {
+                // We can early exit if the digit at `position` is zero.
+                position < self.nonzero_magnitude_end()
+            }
+            RoundingIncrement::MultiplesOf25 => {
+                if position != i16::MAX {
+                    // We can early exit if the digits at `position` and `position + 1` are
+                    // both zeroes.
+                    position + 1 < self.nonzero_magnitude_end()
+                } else {
+                    false
+                }
+            }
+        };
+
+        if self.is_zero() || past_required_digits {
             #[cfg(debug_assertions)]
             self.check_invariants();
             return;
         }
 
         // 3. If the rounding position is *in the middle* of the nonzero digits
-        if magnitude < self.magnitude {
+        if position <= self.magnitude {
             // 3a. Calculate the number of digits to retain and remove the rest
-            let digits_to_retain = crate::ops::i16_abs_sub(self.magnitude, magnitude);
+            let digits_to_retain = crate::ops::i16_abs_sub(self.magnitude, position) + 1;
             self.digits.truncate(digits_to_retain as usize);
-            // 3b. Remove trailing zeros from self.digits to retain invariants
-            // Note: this does not affect visible trailing zeros,
-            // which is tracked by self.lower_magnitude
-            let position_past_last_nonzero_digit = self
+
+            // 3b. Truncate to the previous multiple.
+            match increment {
+                RoundingIncrement::MultiplesOf1 => {
+                    // No need to do more work, trailing zeroes are removed below.
+                }
+                RoundingIncrement::MultiplesOf2 => {
+                    let Some(last_digit) = self.digits.last_mut() else {
+                        // Unreachable
+                        return;
+                    };
+
+                    // Equivalent to (n / 2) * 2, which truncates to the previous
+                    // multiple of two
+                    *last_digit &= 0xFE;
+                }
+                RoundingIncrement::MultiplesOf5 => {
+                    let Some(last_digit) = self.digits.last_mut() else {
+                        // Unreachable
+                        return;
+                    };
+
+                    *last_digit = if *last_digit < 5 { 0 } else { 5 };
+                }
+                RoundingIncrement::MultiplesOf25 => {
+                    // Extend with zeroes to have the correct trailing digits.
+                    self.digits.resize(digits_to_retain as usize, 0);
```

Comment:
Question: what does this do? You already truncated `self.digits` to length `digits_to_retain` up above.

Reply:
This is for cases such as "0.60", which is represented as `digits = [6]`, `magnitude = -1`. Since we need `digits = [6, 0]` instead, we extend to the exact digits of `digits_to_retain` so that we can modify the last, nonexistent zero.

Reply:
I see, okay. "0.3" should truncate down to "0.25" which requires adding an extra digit, and this digit is reflected in `digits_to_retain`.